Locks made transaction termination aware

lutovich opened the pull request:
This PR makes community and forseti locks react on transaction termination by throwing `TransactionTerminatedException`. Previously when a terminated transaction was waiting to grab the lock this waiting did not end.

Feature will allow us to develop kill-transaction functionality for transactions that are basically starved on locks. Feature is currently guarded by a boolean system property 'tx_termination_aware_locks' and off by default.

lutovich committed:
Locks made transaction termination aware

This commit makes community and forseti locks react on transaction termination
by throwing TransactionTerminatedException. Previously when a terminated
transaction was waiting to grab the lock this waiting did not end.

Feature will allow us to develop kill-transaction functionality for
transactions that are basically starved on locks. Feature is currently guarded
by a boolean system property 'tx_termination_aware_locks' and off by default.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -240,16 +245,16 @@ public void failure()
     public boolean shouldBeTerminated()
     {
-        return terminated;
+        return termination.shouldBeTerminated();
```
I'm probably just being silly, but this will now yield a reference on the heap to be traversed every time we check this flag, which every operation on the Kernel API always do. Would it be a bad idea to have KTI implement TxTermination directly?

lutovich replied:
Yeah, this would be an indirection. It is probably not a bad idea for KTI to implement `TxTermination` directly. I'll do this.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/tx/TxTermination.java:
```diff
@@ -0,0 +1,34 @@
+ */
+package org.neo4j.kernel.impl.api.tx;
+
+public interface TxTermination
```
Even for small interfaces and classes it's great to have class level javadoc

lutovich committed:
Made KernelTransaction implement TxTermination

Commit removes TxTermination implementation and makes KernelTransaction
implement this interface directly. This results in less indirections during
`#shouldBeTerminated()` call and fits nicely in KernelTransaction interface
which already had all termination related methods.

lutovich commented:
@tinwelint all comments addressed now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/RWLock.java:
```diff
@@ -545,6 +496,39 @@ public String toString()
+    private void acquireLock( LockType type, Object tx, TxTermination termination, TxLockElement tle )
+    {
+        ragManager.checkWaitOn( this, tx );
+        WaitElement waitElement = new WaitElement( tle, type, currentThread() );
```
it looks like now when you will be interrupted and will try to acquireLock N times you will add WaitElement N times, without calling `ragManager.stopWaitOn( this, tx )`, so wait elements will be accumulated ?

lutovich replied:
`WaitElement`s are cleaned inside `#waitForLock()` if tx was terminated. at the same time `ragManager.checkWaitOn()` modifies internal global map so not calling `ragManager.stopWaitOn()` after tx was terminated is wrong. thanks for spotting this, will fix.

MishaDemianenko commented:
@tinwelint @lutovich guys please be aware that locks in 2.3 and beyond is a bit different in implementation and maybe will require separate pr instead

tinwelint commented:
@MishaDemianenko @lutovich actually transaction termination functionality already exists in 2.3, see `Locks.Client.stop()`. `ForsetiClient` lock acquisition waiting is time based and checks for termination, `CommunityLockClient` could be changed in 2.3 to `wait( aShortWhile )` instead of `wait()` in lock acqusition wait and should be enough. Having that this PR will be a null-merge when forward merging.

If we still want lock acquisition termination like this in 2.2 we should rather backport that functionality from 2.3 than reinventing it here completely.

lutovich commented:
@tinwelint you are right, `Locks.Client.stop()` could be used to terminate transactions waiting on locks with fairly minimal amount of changes. however, backporting that functionality to 2.2 seems like a really big piece of work. it is going to be much more involved than the solution proposed by this PR.

I'd also like to simplify this PR by just including a separate flag in all `Locks.Client` to avoid `TxTermination` all together. Forward merge should definitely be a null-merge.

Any thoughts?

cc @MishaDemianenko

lutovich committed:
Simplify lock client termination

This commit makes lock clients internally know about termination and removes
the need to pass in this info from the owning kernel transaction.

lutovich commented:
Mentioned simplification done. @tinwelint @MishaDemianenko please take a look.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java:
```diff
@@ -102,6 +101,8 @@ public Factory( String key, String... altKeys )
         /** Release all locks. */
         void releaseAll();
 
+        void markForTermination();
```
maybe markAsTerminated?

lutovich replied:
No strong opinion about the name here. `markForTermination` is just consistent with `KernelTransaction#markForTermination()`.

MishaDemianenko replied:
ok, then lets keep it as it is.

MishaDemianenko commented:
@lutovich agree with your statement regarding version 2.3 and above, this pr should be null merge then to those versions and separate PR should created.

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/ForsetiClient.java:
```diff
@@ -656,6 +665,17 @@ public int id()
         return myId;
     }
 
+    private void applyWaitStrategy( Locks.ResourceType resourceType, int tries )
```
This method among others invoked inside of `tryUpgradeToExclusiveWithShareLockHeld`. In case of transaction termination we will end up in generic `catch(Throwable e)` block that will clear wait list and 'throw RuntimeException' as result. Is it appropriate behaviour?

lutovich replied:
This could lead to `TransactionTerminatedException` being wrapped in `RuntimeException`. Fixed in https://github.com/neo4j/neo4j/pull/7168/commits/058fd6cdaf28b65496812256b9b297c0f5406d1c

tkroman commented:
Are there any plans to extend this further to support IO-related operations? I.e. supporting termination during the logical log etc. writing?

MishaDemianenko commented:
@tkroman not in the moment

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/locking/TerminationCompatibility.java:
```diff
@@ -0,0 +1,537 @@
+    private void assertLockAcquisitionSucceeded( LockAcquisition lockAcquisition ) throws Exception
+    {
+        boolean completed = false;
+        for ( int i = 0; i < 20; i++ )
```
20 attempts for 50 ms can be too small for some operation systems

MishaDemianenko commented:
@lutovich another review round completed, couple of comments

lutovich committed:
Fix exception propagation during lock upgrade

Commit fixes the problem where TransactionTerminatedException would be wrapped
in RuntimeException if occurred during shared to exclusive lock upgrade in Forseti.

lutovich committed:
Increase waiting time in TerminationCompatibility

Commit makes this test more reliable by increasing lock wait time
from 1s to 3s for lock related assertions.

MishaDemianenko commented:
@lutovich do not have questions anymore

tinwelint commented:
:+1: looks good to merge
